### ISSUE TITLE
Question: Per route middleware?

### ISSUE THREAD

**Issue by NukeJS** [unjs/h3#177]
Hi,

Is there any way to add middleware per route inside of a router like you can in Express?

**Comment by NozomuIkuta** [comment]
@NukeJS 

Would you provide an example and clarify what you mean by "middleware per route" in Express?

**Comment by NukeJS** [comment]
> @NukeJS
> 
> Would you provide an example and clarify what you mean by "middleware per route" in Express?

```js
function logOriginalUrl (req, res, next) {
  console.log('Request URL:', req.originalUrl)
  next()
}

function logMethod (req, res, next) {
  console.log('Request Type:', req.method)
  next()
}

const logStuff = [logOriginalUrl, logMethod]
app.get('/user/:id', logStuff, (req, res, next) => {
  res.send('User Info')
})
```

Example can be found here: https://expressjs.com/en/guide/using-middleware.html

**Comment by NozomuIkuta** [comment]
I'm still not so sure what you mean even by seeing your code samples, but I guess what you are looking for may be written in [Router section of README](https://github.com/unjs/h3#router). 🤔 

Unlie Express, h3 doesn't provide utility methods whose name corresponds to HTTP methods.

```typescript
  `app[httpMedthod](route, handler)` signature like `app.get(route, hander)`.
Instead, you

**Comment by NukeJS** [comment]
I mean like:

```js
import { createApp, createRouter } from 'h3'

const app = createApp()

const router = createRouter()
 .get('/', [middleware, middleware2], () => 'Hello World!')
 .get('/hello/:name', req => `Hello ${req.context.params.name}!`)

app.use(router)
```

Meaning that as example `middleware` and `middleware2` would only run for the `/` route on GET.

**Comment by NozomuIkuta** [comment]
## TL;DR

This code would work.

```typescript
import { createServer } from 'http'
import { createApp, createRouter, isMethod } from 'h3'

const app = createApp()

app.use([
  {
    match (url, event) {
      if (isMethod(event, 'GET') && url === '/') {
        console.log(`GET request to: '/' (1)`)
        return true
      }
      return false
    },
    handler (event) {
      console.log('Index handler 1')
    }
  },
  {
    match (url, event) {
      if (isMethod(event, 'GET') && url === '/') {
        console.log(`GET request to: '/' (2)`)
        return true
      }
      return false
    },
    handler (event) {
      console.log('Index handler 2')
    }
  }
])

const router = createRouter()
  .get('/', () => 'Hello World!')
  .get('/hello/:name', req => `Hello ${req.context.params.name}!`)

app.use(router)

createServer(app).listen(process.env.PORT || 3000)
```

```bash
curl localhost:3000/
# -> Index handler 1
# -> Index handler 2
# -> Hello World!
```

```bash
curl localhost:3000/hello/nukejs
# -> Hello nukejs!
```

---

## Details


It seems that how route matching works is different between `router.get()` and `app.use()`.

**`router.get()**

- It accepts only one handler at a time, so `router.get([hander1, handler2])` is invalid
- If handler is registered by `router.get()` more than one time, only the last one is used (overwrite)
- Routes registered by `router.get()` are looked up by [unjs/radix3](https://github.com/unjs/radix3)

**`app.use()**

- It accepts a handler, an input layer, or an array of those
- When you pass handler, h3 internally convert it to input layer
- Registered input layers are iterated by for-of loop and handlers of all matched layers are called until either of handlers returns something truthy, all matched handlers are called

## References

- [How handler is registered by `router.http_method_name()`](https://github.com/unjs/h3/blob/1ad516c23cea3d86fca3e820dc80103572d9c2d2/src/router.ts#L22-L45)
- [Definition of InputLayer interface](https://github.com/unjs/h3/blob/1ad516c23cea3d86fca3e820dc80103572d9c2d2/src/app.ts#L24-L33)
- [Input layer iteration in case of `app.use()`](https://github.com/unjs/h3/blob/1ad516c23cea3d86fca3e820dc80103572d9c2d2/src/app.ts#L121-L154)

**Comment by NozomuIkuta** [comment]
More strict speaking, router is also a handler that returns the result of the handler of the matched route, and route matching by router is done within an input layer iteration.

So, the code below may be more simpler to achieve your goal.

```typescript
import { createServer } from 'http'
import { createApp, createRouter, isMethod } from 'h3'

const app = createApp()

const indexHandlerRouter1 = createRouter().get('/', () => console.log('Index handler 1'))
const indexHandlerRouter2 = createRouter().get('/', () => console.log('Index handler 2'))

const router = createRouter()
  .get('/', () => {
    console.log('Index handler 3')
    return 'Index'
  })
  .get('/hello/:name', req => `Hello ${req.context.params.name}!`)

app.use([
  indexHandlerRouter1,
  indexHandlerRouter2,
  router
])

createServer(app).listen(process.env.PORT || 3000)
```

**Comment by mengdu** [comment]
When encountering the same requirement, I solved it this way: 

```ts
export const loginGuard = (h: EventHandler) => async (e: H3Event) => {
  const u = await getLogin(e)
  if (!u) return fail(code.NO_LOGIN)
  return h(e)
}

router.post ('/api/upload', helper.loginGuard (handler.home.updateFile))
```

Expecting official support: 

`router.post ('/api/upload', helper.loginGuard, handler.home.updateFile)` 

or 

`router.post ('/api/upload', [helper.loginGuard, handler.home.updateFile])`

**Comment by pi0** [comment]
Hi. This will be possible via upcoming object syntax. Please follow up via #424

**Comment by clopezpro** [comment]
> Hola. Esto será posible a través de la próxima sintaxis de objetos. Por favor, haga un seguimiento a través del [# 424](https://github.com/unjs/h3/issues/424)



> Hi. This will be possible via upcoming object syntax. Please follow up via #424

thank you hallelujah
